snapcraft/plugins/v2/rust.py

```diff
@@ -0,0 +1,149 @@
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""This rust plugin is useful for building rust based parts.
+
+Rust uses cargo to drive the build.
+
+This plugin uses the common plugin keywords as well as those for "sources".
+For more information check the 'plugins' topic for the former and the
+'sources' topic for the latter.
+
+Additionally, this plugin uses the following plugin-specific keywords:
+
+    - rust-channel
+      (string)
+      select rust channel (stable, beta, nightly)
+
+    - rust-features
+      (list of strings)
+      Features used to build optional dependencies
+
+    - rust-path
+      (list of strings, default [.])
+      Build specific workspace crates
+      Only one item is currently supported.
+
+If a rust-toolchain file is found, the toolchain specified there will
+be used unless rust-channel is set which will override the rust-toolchain
+file. If neither rust-toolchain exists nor rust-channel is set, the latest
+stable toolchain will be used.
+"""
+
+from textwrap import dedent
+from typing import Any, Dict, List, Set
+
+from snapcraft.plugins.v2 import PluginV2
+
+
+class RustPlugin(PluginV2):
+    @classmethod
+    def get_schema(cls) -> Dict[str, Any]:
+        return {
+            "$schema": "http://json-schema.org/draft-04/schema#",
+            "type": "object",
+            "additionalProperties": False,
+            "properties": {
+                "rust-channel": {
+                    "type": "string",
+                    "enum": ["stable", "beta", "nightly"],
+                },
+                "rust-path": {
+                    "type": "array",
+                    "minItems": 1,
+                    # TODO support more than one item.
+                    "maxItems": 1,
+                    "uniqueItems": True,
+                    "items": {"type": "string"},
+                    "default": ["."],
+                },
+                "rust-revision": {"type": "string"},
+                "rust-features": {
+                    "type": "array",
+                    "minItems": 1,
+                    "uniqueItems": True,
+                    "items": {"type": "string"},
+                    "default": [],
+                },
+            },
+            "required": ["source"],
+        }
+
+    def get_build_snaps(self) -> Set[str]:
+        return set()
+
+    def get_build_packages(self) -> Set[str]:
+        return {"curl", "gcc", "git"}
+
+    def get_build_environment(self) -> Dict[str, str]:
+        return {"PATH": "${HOME}/.cargo/bin:${PATH}"}
+
+    def _get_rustup_command(self) -> str:
+        return dedent(
+            """\
+        if ! command -v rustup 2>/dev/null; then
+            curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh -s -- -y --no-modify-path --profile=minimal
```

Comment:
I'm curious what the rationale is for the proto and tls flags? :)

Reply:
That's the recommendation upstream https://rustup.rs/

Reply:
It is indeed, interesting.